### PR TITLE
utils/fork: Check if error pipe is closed

### PR DESCRIPTION
**Pull request by woodruffw**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Mind giving this a stab, @DomT4? I'm away from a Mac at the moment.

### CODE REVIEW

**Comment by DomT4** [comment]
No luck on the manually debugging builds front, I'm afraid. Still bails out hard:

```
^C/usr/local/Homebrew/Library/Homebrew/formula.rb:1847:in `gets'
Interrupt:
1. raise
2. backtrace
3. shell
Choose an action: 1Error: An exception occured within a build process:
  Interrupt:
/usr/local/Homebrew/Library/Homebrew/utils.rb:193:in `wait'
/usr/local/Homebrew/Library/Homebrew/utils.rb:193:in `_system'
/usr/local/Homebrew/Library/Homebrew/utils.rb:199:in `system'
/usr/local/Homebrew/Library/Homebrew/utils.rb:300:in `safe_system'
/usr/local/Homebrew/Library/Homebrew/sandbox.rb:96:in `exec'
/usr/local/Homebrew/Library/Homebrew/formula_installer.rb:735:in `block in build'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:36:in `block (3 levels) in safe_fork'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:30:in `fork'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:30:in `block (2 levels) in safe_fork'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:27:in `open'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:27:in `block in safe_fork'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/tmpdir.rb:89:in `mktmpdir'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:26:in `safe_fork'
/usr/local/Homebrew/Library/Homebrew/formula_installer.rb:719:in `build'
/usr/local/Homebrew/Library/Homebrew/formula_installer.rb:311:in `install'
/usr/local/Homebrew/Library/Homebrew/cmd/install.rb:321:in `install_formula'
/usr/local/Homebrew/Library/Homebrew/cmd/install.rb:253:in `block in install'
/usr/local/Homebrew/Library/Homebrew/cmd/install.rb:251:in `each'
/usr/local/Homebrew/Library/Homebrew/cmd/install.rb:251:in `install'
/usr/local/Homebrew/Library/Homebrew/brew.rb:89:in `<main>'
==> Kept temporary files
Temporary files retained at /private/tmp/libre-20180907-18045-1bxqbci
/usr/local/Homebrew/Library/Homebrew/build.rb:207:in `close': Broken pipe (Errno::EPIPE)
	from /usr/local/Homebrew/Library/Homebrew/build.rb:207:in `rescue in <main>'
	from /usr/local/Homebrew/Library/Homebrew/build.rb:182:in `<main>'
```

And on the build failure front:
```
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:52:in `write': Broken pipe (Errno::EPIPE)
	from /usr/local/Homebrew/Library/Homebrew/utils/fork.rb:52:in `puts'
	from /usr/local/Homebrew/Library/Homebrew/utils/fork.rb:52:in `rescue in block (3 levels) in safe_fork'
	from /usr/local/Homebrew/Library/Homebrew/utils/fork.rb:31:in `block (3 levels) in safe_fork'
	from /usr/local/Homebrew/Library/Homebrew/utils/fork.rb:30:in `fork'
	from /usr/local/Homebrew/Library/Homebrew/utils/fork.rb:30:in `block (2 levels) in safe_fork'
	from /usr/local/Homebrew/Library/Homebrew/utils/fork.rb:27:in `open'
	from /usr/local/Homebrew/Library/Homebrew/utils/fork.rb:27:in `block in safe_fork'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/tmpdir.rb:89:in `mktmpdir'
	from /usr/local/Homebrew/Library/Homebrew/utils/fork.rb:26:in `safe_fork'
	from /usr/local/Homebrew/Library/Homebrew/formula_installer.rb:719:in `build'
	from /usr/local/Homebrew/Library/Homebrew/formula_installer.rb:311:in `install'
	from /usr/local/Homebrew/Library/Homebrew/cmd/install.rb:321:in `install_formula'
	from /usr/local/Homebrew/Library/Homebrew/cmd/install.rb:253:in `block in install'
	from /usr/local/Homebrew/Library/Homebrew/cmd/install.rb:251:in `each'
	from /usr/local/Homebrew/Library/Homebrew/cmd/install.rb:251:in `install'
	from /usr/local/Homebrew/Library/Homebrew/brew.rb:89:in `<main>'
```

**Comment by woodruffw** [comment]
Ugh, I think that closed-pipe issue is being masked by the way we're consuming interrupts from the child process. Moving the interrupt-checking code might do the trick, one sec...

**Comment by DomT4** [comment]
Latest commit resolves the interactive debugging failure but the backtrace on normal build failures remains.

**Comment by woodruffw** [comment]
> but the backtrace on normal build failures remains.

Yep, that makes sense: i meant to check `read.closed?` in https://github.com/Homebrew/brew/pull/4853/commits/997e78673556e9d63181fb04606c6d0767c8e92d but did `write.closed?` instead. I'll fix that, and that *should* fix the `EPIPE` issues.

**Comment by woodruffw** [comment]
Alright, should be ready for another try.

**Comment by DomT4** [comment]
That resolved the normal build failure but reintroduced the interactive debug failure, heh.

```
^C/usr/local/Homebrew/Library/Homebrew/formula.rb:1847:in `gets'
Interrupt:
1. raise
2. backtrace
3. shell
Choose an action: Error: An exception occured within a child process:
  Interrupt:
/usr/local/Homebrew/Library/Homebrew/utils.rb:193:in `wait'
/usr/local/Homebrew/Library/Homebrew/utils.rb:193:in `_system'
/usr/local/Homebrew/Library/Homebrew/utils.rb:199:in `system'
/usr/local/Homebrew/Library/Homebrew/utils.rb:300:in `safe_system'
/usr/local/Homebrew/Library/Homebrew/sandbox.rb:96:in `exec'
/usr/local/Homebrew/Library/Homebrew/formula_installer.rb:735:in `block in build'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:36:in `block (3 levels) in safe_fork'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:30:in `fork'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:30:in `block (2 levels) in safe_fork'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:27:in `open'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:27:in `block in safe_fork'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/tmpdir.rb:89:in `mktmpdir'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:26:in `safe_fork'
/usr/local/Homebrew/Library/Homebrew/formula_installer.rb:719:in `build'
/usr/local/Homebrew/Library/Homebrew/formula_installer.rb:311:in `install'
/usr/local/Homebrew/Library/Homebrew/cmd/install.rb:321:in `install_formula'
/usr/local/Homebrew/Library/Homebrew/cmd/install.rb:253:in `block in install'
/usr/local/Homebrew/Library/Homebrew/cmd/install.rb:251:in `each'
/usr/local/Homebrew/Library/Homebrew/cmd/install.rb:251:in `install'
/usr/local/Homebrew/Library/Homebrew/brew.rb:89:in `<main>'
==> Kept temporary files
Temporary files retained at /private/tmp/libre-20180907-25901-vcbn7y
/usr/local/Homebrew/Library/Homebrew/build.rb:207:in `close': Broken pipe (Errno::EPIPE)
	from /usr/local/Homebrew/Library/Homebrew/build.rb:207:in `rescue in <main>'
	from /usr/local/Homebrew/Library/Homebrew/build.rb:182:in `<main>'
```

Unsure if related but aborting builds part of the way through no longer exits quietly/cleanly:

```
==> make install PREFIX=/usr/local/Cellar/libre/0.5.8
^CError: An exception occured within a child process:
  Interrupt:
```

**Comment by woodruffw** [comment]
Alright, latest should be worth trying again. The root of the problem here is that `build.rb` and `test.rb` install their own exception handlers and exit behavior, while everything else that uses `safe_fork` uses the default handler and exit behavior.

**Comment by DomT4** [comment]
The interactive debug failure remains, sorry 😭

**Comment by woodruffw** [comment]
> The interactive debug failure remains, sorry sob

No problem :slightly_smiling_face:. I added `Interrupt` as one of the child exceptions that we explicitly rewrite, which should catch cases where the error is indeed an interrupt but didn't cause an exit with `130`.

**Comment by DomT4** [comment]
That returns build failures to existing cleanly on abort:
```
  CC      build-x86_64/sipevent/listen.o
  CC      build-x86_64/sipevent/msg.o
  CC      build-x86_64/sipevent/notify.o
^C
```
and makes the interactive debug failure cleaner too, but we seem to land in a situation where you have about a half second to input a choice before it bails out, and if you manage to input a choice _(for example, to drop into the interactive shell in this case)_ this happens:

```
  CC      build-x86_64/sip/via.o
  CC      build-x86_64/sipevent/listen.o
^C/usr/local/Homebrew/Library/Homebrew/formula.rb:1847:in `block in system'
Interrupt:
1. raise
2. backtrace
3. shell
Choose an action: 3
When you exit this shell, you will return to the menu.
%
~>
zsh: error on TTY read: Input/output error
==> Kept temporary files
Temporary files retained at /private/tmp/libre-20180907-35342-cpdgi7
/usr/local/Homebrew/Library/Homebrew/build.rb:207:in `close': Broken pipe (Errno::EPIPE)
	from /usr/local/Homebrew/Library/Homebrew/build.rb:207:in `rescue in <main>'
	from /usr/local/Homebrew/Library/Homebrew/build.rb:182:in `<main>'
```

**Comment by woodruffw** [comment]
Hmm. I think the issue there is either a conflict between the children started by `interactive_shell`/`safe_fork` *or* an incorrect assumption about our state when we're resuming execution after an interrupt (or maybe both?). It's getting late here, so I'll do some more investigating tomorrow.

**Comment by DomT4** [comment]
No worries. It's slightly annoying but I don't think it's a _"revert urgently"_ situation, so I'm personally okay with leaving it be for tonight to give you time to look at it tomorrow without having to reapply the whole changeset.

File: Library/Homebrew/utils/fork.rb [hunk 1]
```diff
@@ -46,8 +48,15 @@ def self.safe_fork(&_block)
               error_hash["output"] = e.output
             end
 
-            write.puts error_hash.to_json
-            write.close
+            # Attempt to write our exception back to the parent, failing silently
+            # if the pipe if broken (which would indicate that we've already written an
```
Comment (reitermarkus) [inline Library/Homebrew/utils/fork.rb]: is broken

**Comment by woodruffw** [comment]
One of the subtle changes in #4819 is that we no longer read until `EWOULDBLOCK`, but only until the first JSON-serialized exception is available. I just reverted that behavior, which should prevent the `EPIPE` issues entirely. It remains to be seen whether that'll regress the interrupt issues, though...

File: Library/Homebrew/utils/fork.rb [hunk 2]
```diff
@@ -64,21 +67,20 @@ def self.safe_fork(&_block)
...
           end
           write.close
-          # Each line on the error pipe contains a JSON-serialized exception.
-          # We read the first, since only one is necessary for a failure.
-          data = read.gets
+          data = read.read
           read.close
           Process.wait(pid) unless socket.nil?
 
+          raise Interrupt if $CHILD_STATUS.exitstatus == 130
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/fork.rb]: Could you add a comment explaining why the `130` here? I notice that it's code that's moved from below but if you have context now it'd be great.

File: Library/Homebrew/utils/fork.rb [hunk 3]
```diff
@@ -64,21 +67,20 @@ def self.safe_fork(&_block)
...
+
           if data && !data.empty?
-            error_hash = JSON.parse(data) unless data.nil? || data.empty?
+            error_hash = JSON.parse(data.lines.first)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/fork.rb]: Hooray for the death of another `unless ... || ...`!

**Comment by woodruffw** [comment]
> Will this PR address my error?

It should, yeah.

I'll address @MikeMcQuaid's documentation request before merging tonight, but otherwise this should be good to go.

**Comment by commitay** [comment]
@woodruffw gentle ping on this.

**Comment by woodruffw** [comment]
Sorry, I've been a little swamped lately. I *think* this should be good to merge as-is; let me just address the comment above.

**Pull request merged**